server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -572,41 +572,28 @@ public HandshakeResponse(StreamInput in) throws IOException {
                 // message, but recognise that this may fail
                 discoveryNode = new DiscoveryNode(in);
             } catch (Exception e) {
-                if (isIncompatibleBuild(version, buildHash)) {
-                    throw new IllegalArgumentException(
-                        "unidentifiable remote node is build ["
-                            + buildHash
-                            + "] of version ["
-                            + version
-                            + "] but this node is build ["
-                            + Build.CURRENT.hash()
-                            + "] of version ["
-                            + Version.CURRENT
-                            + "] which has an incompatible wire format",
-                        e
-                    );
-                } else {
-                    throw e;
-                }
+                failIfIncompatibleBuild(e);
+                throw e;
             }
+            failIfIncompatibleBuild(null);
+            clusterName = new ClusterName(in);
+        }
 
+        private void failIfIncompatibleBuild(@Nullable Exception cause) {
             if (isIncompatibleBuild(version, buildHash)) {
                 throw new IllegalArgumentException(
-                    "remote node ["
-                        + discoveryNode
-                        + "] is build ["
+                    "unidentifiable remote node is build ["
```

Comment:
I think we need to keep the `remote node [...]` bit in cases where the node isn't unidentifiable. That's almost always the path we use and without that it's almost impossible to determine which node is the problem.